examples/specific/array.h

```diff
@@ -3,5 +3,4 @@
 int foo(int a[5]);
 
 /** My other function */
-int bar(int n, int a[static n]);
-
+int bar(int n, int a[]);
```

Comment:
Revert. This is supposed to be a C example.

Reply:
My VSCode spits out an error about this
```
expected an expressionC/C++(29)
```
I'll revert it if the original syntax was intended. BTW I've never seen such a syntax.

Reply:
After reverting this Sphinx throws a warning:
```
sphinx.errors.SphinxWarning: /path/to/breathe/documentation/source/specific.rst.rst:120:Parsing of expression failed. Using fallback parser. Error was:
Error in postfix expression, expected primary expression or type.
If primary expression:
  Invalid C++ declaration: Expected identifier in nested name, got keyword: static [error at 28]
    int bar (int n, int a[static n])
    ----------------------------^
If type:
  Invalid C++ declaration: Expected identifier in nested name, got keyword: static [error at 28]
    int bar (int n, int a[static n])
    ----------------------------^
```

Reply:
I could not fix this by setting `OPTIMIZE_OUTPUT_FOR_C = YES` in the project's Doxyfile. Sphinx still thinks that it is a C++ declaration (for some reason).

Reply:
Tagging @michaeljones because git names him as the initial author of this file. Is the syntax of this declaration important to the test it is used for?

The XML for the test project seems to always be flagged as "C++". 
```xml
  <compounddef id="array_8h" kind="file" language="C++">
```
This is despite what extension I use (tried renaming the file to `array.c`). I also tried 
```ini
EXTENSION_MAPPING = h=C
# or
EXTENSION_MAPPING = "h=C"
# while also setting
OPTIMIZE_OUTPUT_FOR_C = YES
```
But, no matter what, the XML output is telling sphinx that this syntax is C++ (which results in a warning about a malformed declaration).

Even, manually changing the language specified in XML to `language="C"` has no effect toward sphinx (still thinks its C++ syntax).

I see the conf.py has specified
```py
breathe_domain_by_file_pattern = {
    "class.h": "cpp",
    "alias.h": "c",
    "array.h": "c",
    "c_*.h": "c",
}
```
but `"array.h": "c"` doesn't seem to be respected in this case.

Reply:
Strange. In my local build it is indeed translated to a ``.. c:function::``as it should (instead of ``.. cpp:function::``).
If you in ``conf.py`` set ``breathe_debug_trace_directives = True`` then you can see an indication of which directives are sent to Sphinx, more directly.

As you probably found, C arrays is one of the few cases where C is larger than C++, but for reference: https://en.cppreference.com/w/c/language/array

Reply:
And I locally apparently use Doxygen 1.9.2.

Reply:
Thanks that was my next question.

With debug_trace asserted, I found that it is indeed is using
```
Running directive: .. c:function::  int bar (int n, int a[static n])
```
# 🤷🏼 
I'm running Doxygen v1.9.3 locally.

Reply:
I'm afraid I can only contribute by saying that I don't know the context of the example any more. I don't know what may or maybe not be important.

Reply:
FWIW, I'm locally using Sphinx v4.4.0 (like the CI does).

Reply:
I'm confused now: the trace says ``.. c:function::`` but the error says ``Invalid C++ declaration``?

Reply:
I still get the same error with Doxygen v1.9.2

Reply:
In hopes that this was just my setup I did revert this change, but it turns out the error is reproduced in the CI workflow.

I'm starting to feel like this is an new issue we just didn't know about because all warnings were ignored. I'd raise the issue on a separate thread, but I can't narrow down the cause.

Reply:
I've tried using
- Doxygen v1.9.3 + Sphinx v4.4.0
- Doxygen v1.8.17 + Sphinx v4.4.0
- Doxygen v1.8.17 + Sphinx v3.0.0

All combinations are raising this warning. If I ignore this warning the generated HTML for this function is
```html
<dt class="sig sig-object c" id="c.bar">
```

Reply:
I elevated this problem to #854 for proper tracking.

For now I've made the docs pass by removing the `static n` snippet, but this should be added back in when we resolve #854 .